ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -180,6 +182,37 @@ public void appendFrom(ReadableByteChannel channel, long size) throws IOExceptio
     }
   }
 
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Attempts to write the {@code byteArray} to this segment in direct IO manner.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param byteArray The bytes from which data needs to be written from.
+   * @param offset The offset in the byteArray to start with.
+   * @param length The amount of data in bytes to use from the byteArray.
+   * @throws IllegalArgumentException if there is not enough space for data of size {@code length}.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  public int appendFromDirectly(byte[] byteArray, int offset, int length) throws IOException {
+    if (!fileChannel.isOpen()) {
+      throw new ClosedChannelException();
+    }
+    if (endOffset.get() + length > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
+          "Channel cannot be written to segment [" + file.getAbsolutePath() + "] because" + " it exceeds the capacity ["
+              + capacityInBytes + "]");
+    } else {
+      try (DirectRandomAccessFile directFile = new DirectRandomAccessFile(file, "rw", 2 * 1024 * 1024)) {
```

Comment:
is opening up the file on every write a good idea? Is it better to open it the first time this function is called (or in the constructor) and use it only in this function (and close() it on close)

Just like `initBuffer()` and `dropBuffer()` in #998, we can have functions that init and close the direct IO file (only compaction uses it so maybe it is something that the log opened up by compaction can do)

Reply:
It's not a good idea to open on every write. Will address this issue in a follow up PR.